File: Components/Button.py
from Component import Component
from Enums import ComponentType, AnswerType
from QA import Question, Answer

class Button(Component):
    def __init__(self, id, board):
        super().__init__(id, board)
        self.name = "button"
        self.device_type = ComponentType.INPUT_DEVICE_ONLY_DIGITAL
        self.question = Question(self.init, "mode", "What kind of input does the button take?",
                                 AnswerType.MULTI_OPTION,
                                 [
                                     Answer("when button is pressed for once", "press"),
                                     Answer("when button is held down", "held"),
                                     Answer("when button is not pressed", "not")
                                 ]
                                 )
        self._pin = "button" + str(id) + "_pin"
        self._val = "button" + str(id) + "_val"
        self._prev = "button" + str(id) + "_prev"
        
    def get_global_var(self):
        ret = self.str_init_variable("const int", self._pin, '2') #TEMP
        ret += self.str_init_variable("int", self._val, '0')
        if self.init["mode"] is "press":
            ret += self.str_init_variable("int", self._prev, '0')
        return ret
    
    def get_setup(self):
        return self.str_pinMode(self._pin, 'o')
    
    def get_loop_start(self):
        ret = ''
        if self.init["mode"] == "press":
            ret += self.str_assign_variable(self._prev, self._val)
        ret += self.str_assign_variable(self._val, 'digitalRead(' + self._pin + ');')
        return ret
    
    def get_loop_logic(self, state_num = 0):
        print(self.init)
        print(self.states[state_num])
        print(self.init["mode"])
        match self.init["mode"]:
            case "press":
                ret = self._prev + " == 0 &&" + self._val + " == 1"
            case "held":
                ret = self._val + " == 1"
            case "not":
                ret = self._val + " == 0"
            case _:
                ret = ""
            
        return ret

File: codeplate/src/Arduino_Components/Button.js
import React from "react";
import Component from "./Components.js";
import { ComponentType, AnswerType, STAGE } from "./Tools/Enums.js";
import { Question, Answer } from "./Tools/QA.js";
import Code from "./Tools/Code.js";

class Button extends Component {
    constructor(props){
        super(props)
        this.state = {
            deviceType: ComponentType.INPUT_DEVICE_ONLY_DIGITAL,
            init: 0,
            question: <Question handleAnswer = {this.updateAnswer}
                                    questionText = "When should the button trigger an output?"
                                    answerType = {AnswerType.MULTI_OPTION}
                                    answerOption = {
                                        [
                                            {
                                                text: "when the button is pressed briefly",
                                                value: "press",
                                                followup: ""
                                            },
                                            {
                                                text: "when the button is held down",
                                                value: "held",
                                                followup: ""
                                            },
                                            {
                                                text: "when the button is not being held down",
                                                value: "not",
                                                followup: ""
                                            }
                                        ]
                                    }
                                    />,
            _pin: "button" + props.id + "_pin",
            _val: "button" + props.id + "_val",
            _prev: "button" + props.id + "_prev",
            code: new Code()
        }
    }

    getName = () => { return "Button" }

    updateAnswer = (answer, option) => {
        this.setState({mode: answer})
        this.props.handlePropsChange({mode: answer}, this.props.id, "INPUT")
        this.props.handleCode("INPUT", this.props.id, this.getGlobalVar(answer), this.getSetup(), this.getLoopStart(answer), this.getLoopLogic(answer), this.getHelperFunction())
    }

    getGlobalVar = (mode) => {
        var ret = [this.state.code.strInitVariable("const int", this.state._pin, '2'), this.state.code.strInitVariable("int", this.state._val, '0')]
        if (mode === "press")
            ret.push(this.state.code.strInitVariable("int", this.state._prev, '0'))
        return ret
    }
    getSetup = () => {
        return [this.state.code.strPinMode(this.state._pin, 'o')]
    }
    getLoopStart = (mode) => {
        var ret = []
        if (mode == "press")
            ret.push(this.state.code.strAssignVariable(this.state._prev, this.state._val))
        ret.push(this.state.code.strAssignVariable(this.state._val, 'digitalRead(' + this.state._pin + ')'))
        return ret
    } //TODO
    getLoopLogic = (mode) => {
        var ret = []
        console.log(mode)
        switch (mode){
            case "press":
                ret.push(this.state._prev + " == 0 &&" + this.state._val + " == 1")
            case "held":
                ret.push(this.state._val + " == 1")
            case "not":
                ret.push(this.state._val + " == 0")
        }
        return ret
    }
    getHelperFunction = () => {
        return []
    }

    render() {
        if (this.props.getStage() === STAGE.RENDER_CODE) {
            return (
                <>
                </>
            )
        }
        return (
            <div>
                {this.state.question}
            </div>
        );
    }
}

export default Button;
